### PR TITLE
refactor(components): [slider] switch to script-setup syntax

### PR DESCRIPTION
**Pull request by holazz**
https://github.com/element-plus/element-plus/issues/6114

- switch to `setup`, export instance type, use `withInstall`
- extract props, emits, tokens and hooks
- `index.vue` -> `slider.vue`, `marker.vue` ->`marker.ts` 
- improve types

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by holazz** [comment]
cc @sxzz @jw-foss

File: packages/components/slider/src/slider.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  firstValue: number
+  secondValue: number
+  oldValue?: number | number[]
+  dragging: boolean
+  sliderSize: number
+}
+
+export const sliderProps = buildProps({
+  modelValue: {
+    type: definePropType<number | number[]>([Number, Array]),
```
Comment (sxzz) [inline packages/components/slider/src/slider.ts:30]:
```suggestion
    type: definePropType<Arrayable<number>>([Number, Array]),
```

File: packages/components/slider/src/composables/useSlide.ts [hunk 2]
```diff
@@ -5,28 +5,24 @@ import {
...
 
 export const useSlide = (
-  props: ISliderProps,
-  initData: ISliderInitData,
-  emit: ComponentInternalInstance['emit']
+  props: SliderProps,
+  initData: SliderInitData,
+  emit: SetupContext<SliderEmits>['emit']
 ) => {
   const elForm = inject(formContextKey, {} as FormContext)
```
Comment (sxzz) [inline packages/components/slider/src/composables/useSlide.ts:18]: Please use `useFormItem`

File: packages/components/slider/src/slider.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    type: String,
+    default: undefined,
+  },
+  marks: {
+    type: definePropType<SliderMarks>(Object),
+  },
+} as const)
+export type SliderProps = ExtractPropTypes<typeof sliderProps>
+
+const emitChangeFn = (value: Arrayable<number>) =>
```
Comment (sxzz) [inline packages/components/slider/src/slider.ts:99]:
```suggestion
const isValidValue = (value: Arrayable<number>) =>
```

**Review by sxzz — COMMENTED** [review]
Please rename filenames of composables to `kebab-case`.

**Comment by holazz** [comment]
@sxzz all resolved.

**Comment by holazz** [comment]
> @sxzz all resolved.

@sxzz There is a problem of heap out of memory when building, I am still looking for the reason 🤔

**Comment by holazz** [comment]
Hi, @sxzz! This problem seems to be related to https://github.com/element-plus/element-plus/pull/8018. Before this commit, I was able to build successfully using the same code. After this, there is a problem of heap out of memory during the build. And I found that the type file after the build has twice as many lines of code as before. But I still haven't figured out the direct cause of this problem. What do you think about this?

before

after

**Comment by holazz** [comment]
I tried to refactor the `el-tooltip` to script syntax on the latest commit, and the same build problem also occurred.

https://elementplus-elementplus-5ezxokuw28m.ws-us46.gitpod.io/

**Comment by holazz** [comment]
Hi, @sxzz! I use p-limit to make Promises execute sequentially. I found a ref of the component instance type (eg `ref<SliderButtonInstance>()`) is defined in every file that fails the build. Then I replaced all `ref<.*>` with `ref<any>` and built success!

https://elementplus-elementplus-i24tzhp418p.ws-us46.gitpod.io/

**Comment by holazz** [comment]
> @holazz You can try to rebase to latest `dev` branch now! The issue has been solved.

Great! I was just curious in which commit this issue was resolved. 🤔

**Comment by holazz** [comment]
> @holazz
> 
> * [80c6eeb#diff-ed6243f9620b721d2465aa81cb41775342fd581ad67a274a52406b54eebbfc74](https://github.com/element-plus/element-plus/commit/80c6eebd126fd20913167328668c88bdf85c818e#diff-ed6243f9620b721d2465aa81cb41775342fd581ad67a274a52406b54eebbfc74)
> * [ci: fix build staging #8143](https://github.com/element-plus/element-plus/pull/8143)

Thanks, I also tried but it didn't work, maybe I'm not using it correctly. 😅

**Review by sxzz — APPROVED** [review]

**Pull request merged**